glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/DeletionPingUploadWorker.kt

```diff
@@ -0,0 +1,192 @@
+        // In particular, there are issues if the pings are cleared (as part of
+        // disabling telemetry), while the ping uploader is trying to upload queued pings.
+        // Therefore, this lock is held both when uploading pings and when calling
+        // into the Rust code that might clear queued pings (set_upload_enabled).
+        internal val pingQueueLock = Any()
+
+        /**
+         * Build the constraints around which the worker can be run, such as whether network
+         * connectivity is required.
+         *
+         * @return [Constraints] object containing the required work constraints
+         */
+        @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+        fun buildConstraints(): Constraints = Constraints.Builder()
+            .setRequiredNetworkType(NetworkType.CONNECTED)
+            .build()
+
+        /**
+         * Build the [OneTimeWorkRequest] for enqueueing in the [WorkManager].  This also adds a tag
+         * by which enqueued requests can be identified.
+         *
+         * @return [OneTimeWorkRequest] representing the task for the [WorkManager] to enqueue and run
+         */
+        internal fun buildWorkRequest(): OneTimeWorkRequest = OneTimeWorkRequestBuilder<DeletionPingUploadWorker>()
+            .addTag(PING_WORKER_TAG)
+            .setConstraints(buildConstraints())
+            .build()
+
+        /**
+         * Function to aid in properly enqueuing the worker in [WorkManager]
+         *
+         * @param context the application [Context] to get the [WorkManager] instance for
+         */
+        internal fun enqueueWorker(context: Context) {
+            WorkManager.getInstance(context).enqueueUniqueWork(
+                PING_WORKER_TAG,
+                ExistingWorkPolicy.KEEP,
+                buildWorkRequest())
+        }
+
+        /**
+         * Function to perform the actual ping upload task.  This is created here in the
+         * companion object in order to facilitate testing.
+         *
+         * @return true if process was successful
+         */
+        internal fun uploadPings(): Boolean = process()
+
+        /**
+         * Function to cancel any pending ping upload workers
+         *
+         * @param context the application [Context] to get the [WorkManager] instance for
+         */
+        internal fun cancel(context: Context) {
+            WorkManager.getInstance(context).cancelUniqueWork(PING_WORKER_TAG)
+        }
+
+        /**
+         * Function to deserialize and process all serialized ping files.  This function will ignore
+         * files that don't match the UUID regex and just delete them to prevent files from polluting
+         * the ping storage directory.
+         *
+         * @return Boolean representing the success of the upload task. This may be the value bubbled up
+         *         from the callback, or if there was an error reading the files.
+         */
+        fun process(): Boolean {
```

Comment:
Any chance we could just refactor out all this code or use the one from the main uploader?